internal/consensus/reactor.go

```diff
@@ -1003,6 +778,198 @@ OUTER_LOOP:
 	}
 }
 
+// pick a block part to send if the peer has the same part set header as us or if they're catching up and we have the block.
+// returns the part and a bool that signals whether to continue to the loop (true) or to sleep.
+// NOTE there is one case where we don't return a part but continue the loop (ie. we return (nil, true)).
+func pickPartToSend(
+	logger log.Logger,
+	blockStore sm.BlockStore,
+	rs *cstypes.RoundState,
+	ps *PeerState,
+	prs *cstypes.PeerRoundState,
+) (*types.Part, bool) {
+	// If peer has same part set header as us, send block parts
+	if rs.ProposalBlockParts.HasHeader(prs.ProposalBlockPartSetHeader) {
+		if index, ok := rs.ProposalBlockParts.BitArray().Sub(prs.ProposalBlockParts.Copy()).PickRandom(); ok {
+			part := rs.ProposalBlockParts.GetPart(index)
+			return part, true
+		}
+	}
+
+	// If the peer is on a previous height that we have, help catch up.
+	blockStoreBase := blockStore.Base()
+	if blockStoreBase > 0 &&
+		0 < prs.Height && prs.Height < rs.Height &&
+		prs.Height >= blockStoreBase {
+		heightLogger := logger.With("height", prs.Height)
+
+		// if we never received the commit message from the peer, the block parts won't be initialized
+		if prs.ProposalBlockParts == nil {
+			blockMeta := blockStore.LoadBlockMeta(prs.Height)
+			if blockMeta == nil {
+				heightLogger.Error("Failed to load block meta",
+					"blockstoreBase", blockStoreBase, "blockstoreHeight", blockStore.Height())
+				return nil, false
```

Comment:
In the original code, we sleep here:

time.Sleep(conR.conS.config.PeerGossipSleepDuration)

then we continue to the OUTER_LOOP.

Reply:
In summary, the operation is different.

Reply:
https://github.com/cometbft/cometbft/blob/dc4590e1c874eda7d760d14e39f656b3e2622cdf/internal/consensus/reactor.go#L617-L628

Reply:
So, correcting the rationale above. The original code sleeps in this case before restarting the main for loop. In this refactored version, this is implemented by not continuing (`continueLoop = false`) and sleeping at the end of the main for loop.

In other words, this behavior is equivalent to the original code's.

Reply:
Shall we resolve this?